stan/math/gpu/cholesky_decompose.hpp

```diff
@@ -0,0 +1,127 @@
+                                               const int block = 100,
+                                               const int divider = 2,
+                                               const int min_block = 100) {
+  matrix_gpu L(A.rows(), A.cols());
+  if (block <= min_block || divider <= 1) {
+    try {
+      opencl_kernels::cholesky_decompose(cl::NDRange(A.rows()),
+                                         cl::NDRange(A.rows()), A.buffer(),
+                                         L.buffer(), A.rows());
+    } catch (const cl::Error& e) {
+      check_opencl_error("cholesky_decompose", e);
+    }
+  } else {
+    L = stan::math::cholesky_decompose(A, block / divider);
+  }
+  return L;
+}
+}  // namespace internal
+/**
+ * Return the lower-triangular Cholesky factor (i.e., matrix
+ * square root) of the specified square, symmetric matrix.
+ * The return value \f$L\f$ will be a lower-traingular matrix such that the
+ * original matrix \f$A\f$ is given by
+ * <p>\f$A = L \times L^T\f$.
+ * The Cholesky decomposition is computed on the GPU. This algorithm is
+ * recursive, where The parameters <code>block</code>, <code>divider</code>, and
+ *  <code>min_block</code> act as tuning parameters for the recursive step of
+ *  the GPU based Cholesky decompostion. The matrix is subset by the
+ *  <code>block</code> size, and if the <code>block</code> size is less than
+ * <code>min_block</code> then the cholesky decomposition on the GPU is computed
+ * using that submatrix. If <code>block</code> is greater than
+ * <code>block_size</code> then <code>cholesky_decompose</code> is run again
+ * with <code>block</code> equal to <code>block/divider</code>. Once the
+ * Cholesky Decomposition is computed, the full matrix cholesky is created
+ * by propogating the cholesky forward as given in the reference report below.
+ *
+ * For a full guide to how this works
+ * see the Cholesy decompostion chapter in the  reference report
+ * <a href="https://goo.gl/6kWkJ5"> here</a>.
+ * @param A Symmetric matrix on the GPU.
+ * @param block Size of the block used to compute the cholesky decomposition.
+ * @param divider Proportion to divide the submatrix by at each recursive step.
+ * @param min_block The amount that block is checked against to decide
+ *  whether to continue the recursion or to perform the cholesky.
+ * @return Square root of matrix on the GPU.
+ * @throw std::domain_error if m is not
+ *  positive definite (if m has more than 0 elements)
+ */
+inline matrix_gpu cholesky_decompose(matrix_gpu& A, const int block,
+                                     const int divider, const int min_block) {
+  auto offset = 0;
+  // NOTE: The code in this section follows the naming conventions
+  // in the report linked in the docs.
+  matrix_gpu A_11(block, block);
+  // Repeats the blocked cholesky decomposition until the size of the remaining
+  // submatrix is smaller or equal to the block size
+  while ((offset + block) < (A.rows())) {
```

Comment:
I think I'm confused about how the recursion and while loop are interacting, sorry! Will try to get to the bottom of it.

Reply:
Yeah, I need to cleanup this recursion the way you suggest. That might make it more clear. Sorry about that.

Reply:
It doesn't help that I can't run this code locally... :'(

Reply:
Oh no worries! I am not totally sure the way I suggested is right... trying to get something running locally to try things out.

Reply:
>It doesn't help that I can't run this code locally... :'(

Double frowny face, I'll try to run this on my mac tmrw and figure out what I need to do to get it working. Are you on Mojave?

Reply:
>  I'm not sure what this would look like using the OpenCL API - it might rely on the Async framework @SteveBronder was telling me about, or it might also be implemented at the kernel level (I think?)

I'm not seeing how we can do this completely in parallel since A_11 depends on what was computed and placed in A during the previous iteration.

But with OpenCL 2.0 we could call all of this on the device side (i.e. when one kernel finishes it calls the next kernel without the host having to do anything). This would give these sorts of problems quite a lot of speedup. If @rok-cesnovar want's to clean this up then that's cool and good but otherwise I think we can hold off on making this change.